src/shared/Microsoft.Git.CredentialManager/Authentication/OAuth/OAuth2Client.cs

```diff
@@ -0,0 +1,440 @@
+            {
+                string json = await response.Content.ReadAsStringAsync();
+
+                if (response.IsSuccessStatusCode && TryCreateTokenEndpointResult(json, out OAuth2TokenResult result))
+                {
+                    return result;
+                }
+
+                throw CreateExceptionFromResponse(json);
+            }
+        }
+
+        public async Task<OAuth2TokenResult> GetTokenByDeviceCodeAsync(OAuth2DeviceCodeResult deviceCodeResult)
+        {
+            var formData = new Dictionary<string, string>
+            {
+                [OAuth2Constants.DeviceAuthorization.GrantTypeParameter] = OAuth2Constants.DeviceAuthorization.DeviceCodeGrantType,
+                [OAuth2Constants.DeviceAuthorization.DeviceCodeParameter] = deviceCodeResult.DeviceCode,
+                [OAuth2Constants.ClientIdParameter] = _clientId,
+            };
+
+            TimeSpan retryInterval = deviceCodeResult.PollingInterval;
+            while (true)
+            {
+                try
+                {
+                    using (HttpContent requestContent = new FormUrlEncodedContent(formData))
+                    using (HttpRequestMessage request = CreateRequestMessage(HttpMethod.Post, _endpoints.TokenEndpoint, requestContent))
+                    using (HttpResponseMessage response = await _httpClient.SendAsync(request))
+                    {
+                        string json = await response.Content.ReadAsStringAsync();
+
+                        if (response.IsSuccessStatusCode && TryCreateTokenEndpointResult(json, out OAuth2TokenResult result))
+                        {
+                            return result;
+                        }
+
+                        var error = JsonConvert.DeserializeObject<ErrorResponseJson>(json);
+
+                        switch (error.Error)
+                        {
+                            case OAuth2Constants.DeviceAuthorization.Errors.AuthorizationPending:
+                                // Retry with the current polling interval value
+                                break;
+                            case OAuth2Constants.DeviceAuthorization.Errors.SlowDown:
+                                // We must increase the polling interval by 5 seconds
+                                retryInterval = retryInterval.Add(TimeSpan.FromSeconds(5));
+                                break;
+                            default:
+                                // For all other errors do not retry
+                                throw CreateExceptionFromResponse(json);
+                        }
+                    }
+                }
+                catch (TimeoutException)
+                {
+                    // Back-off exponentially (2 * x = x + x)
+                    retryInterval += retryInterval;
+                }
+
+                // Wait the polling interval before retrying
+                await Task.Delay(retryInterval);

```

Comment:
This has no stop case if the user cannot finish the device flow. Should it?

Reply:
Added cancellation token support.